provider/mediaconvert/mediaconvert.go

```diff
@@ -0,0 +1,376 @@
+package mediaconvert
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/NYTimes/video-transcoding-api/config"
+	"github.com/NYTimes/video-transcoding-api/db"
+	"github.com/NYTimes/video-transcoding-api/provider"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/external"
+	"github.com/aws/aws-sdk-go-v2/service/mediaconvert"
+	"github.com/pkg/errors"
+)
+
+const (
+	// Name identifies the MediaConvert provider by name
+	Name = "mediaconvert"
+
+	defaultAudioSampleRate = 48000
+)
+
+func init() {
+	_ = provider.Register(Name, mediaconvertFactory)
+}
+
+type mediaconvertClient interface {
+	CreateJobRequest(*mediaconvert.CreateJobInput) mediaconvert.CreateJobRequest
+	GetJobRequest(*mediaconvert.GetJobInput) mediaconvert.GetJobRequest
+	ListJobsRequest(*mediaconvert.ListJobsInput) mediaconvert.ListJobsRequest
+	CancelJobRequest(*mediaconvert.CancelJobInput) mediaconvert.CancelJobRequest
+	CreatePresetRequest(*mediaconvert.CreatePresetInput) mediaconvert.CreatePresetRequest
+	GetPresetRequest(*mediaconvert.GetPresetInput) mediaconvert.GetPresetRequest
+	DeletePresetRequest(*mediaconvert.DeletePresetInput) mediaconvert.DeletePresetRequest
+}
+
+type mcProvider struct {
+	client mediaconvertClient
+	cfg    *config.MediaConvert
+}
+
+func (p *mcProvider) Transcode(job *db.Job) (*provider.JobStatus, error) {
+	presetCh := make(chan *presetResult)
+	presets := map[string]mediaconvert.Preset{}
+
+	var wg sync.WaitGroup
+	for _, output := range job.Outputs {
+		presetID, ok := output.Preset.ProviderMapping[Name]
+		if !ok {
+			return nil, provider.ErrPresetMapNotFound
+		}
+
+		wg.Add(1)
+		go p.makeGetPresetRequest(presetID, presetCh, &wg)
+	}
+
+	go func() {
+		wg.Wait()
+		close(presetCh)
+	}()
+
+	for res := range presetCh {
+		if res.err != nil {
+			return nil, fmt.Errorf("error getting preset info: %s", res.err)
+		}
+
+		presets[res.presetID] = res.preset
+	}
```

Comment:
Could we abstract all of that into a function that returns the presets map? Something like:

```go
func (p *mcProvider) getPresets(job *db.Job) (map[string]mediaconvert.Preset, error) {
    // impl...
}
```

Reply:
Absolutely, I'll do this.

Reply:
done